weaknesses/MASVS-CRYPTO/MASWE-0010.md

```diff
@@ -14,9 +14,39 @@ refs:
+
+- If the password is smaller than the key, the full key space isn't used. The remaining space is padded (spaces are sometimes used for padding).
+- A user-supplied password will realistically consist mostly of displayable and pronounceable characters. Therefore, only some of the possible 256 ASCII characters are used and entropy is decreased by approximately a factor of four.
+
+## Impact
+
+- **Risk of Brute-Force Attacks**: Improper Key derivation functions open up for brute force password- and secret cracking, key or dictionary attacks such as rainbow tables.
+- **Loss of Confidentiality**: Improper Key derivation may allow an attacker to guess or find the input and therefore steal the user's password or cryptographic key.
+- **Loss of Integrity**: Given that the attacker has access to the user's password or cryptographic key, the overall security of the app and mobile phone may be compromised.
+
+## Modes of Introduction
+
+- **Using static or predictable salt on low-entropy input**: Using user-supplied input together with a static or predictable salt makes it possible for the attacker to to pre-compute the hash value using dictionary attack, effectively disabling the protection that a salt would provide.
+- **Using user-supplied salt**: Using user-supplied salt will make it possible for an attacker to extract all derived keys through a KDF by supplying the user-supplied salt as an argument.
+- **Using non-resource intensive algorithms on low-entropy input**: Using user-supplied input together with HKDF makes it easy for password crackers to execute a preimage attack.
+- **Use a hash function as a general-purpose KDF**: In scenarios where the information used during key derivation is attacker-controlled, using a integrity based hash function (e.g from the SHA family) as KDF can expose the application to brute force or length-extension attacks.
+- **Using Deprecated, Risky or Broken Algorithms**: Relying on deprecated, risky or inherently broken cryptographic algorithms can result in the generation of weaker keys. As these algorithms often have vulnerabilities or support shorter key lengths, they are more susceptible to modern attacks, compromising the overall security of the app.
+
+## Mitigations
+
+- **Generate random salt using CSPRNGs with high entropy seeding**: Ensuring the use of strong, cryptographically secure PRNGs called CSPRNGs with high entropy seeding is essential for robust key security.
+- **Use Recommended and Approved algorithms that are fit for purpose**: In cases where the input is user-controlled, use key derivation functions such as Argon2, scrypt, bcrypt or PBKDF2 that provides a sufficient level of computational effort. Otherwise, ensure the input is thoroughly random using a recommended CSPRNG that guarantees high entropy seeding.
```

Comment:
```suggestion
- **Use Recommended and Approved algorithms that are fit for purpose**: In cases where the input is user-controlled, use key derivation functions such as Argon2, scrypt, bcrypt or PBKDF2 that provide a sufficient level of computational effort. Otherwise, ensure the input is thoroughly random using a recommended CSPRNG that guarantees high entropy seeding.
```
There should also be a general mitigation about choosing a better architecture where the encryption/hashing is done serverside or using the device's pin. I've seen many locally implemented pincode encryption schemes which are all vulnerable to a bruteforce once the cryptographic material is obtained.